### ISSUE TITLE
Client memory corruption when parsing server response (Segmentation fault, KVIrc v5.0.0)

### ISSUE THREAD

**Issue by ghost** [kvirc/KVIrc#2452]
If while connected to a server, the server sends the following string ":C PRIVMSG    ", the client segfaults.
```

Core was generated by `kvirc'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007ff8fcf70096 in KviCString::KviCString (this=0x7ffc13d31290, str=...) at /home/wssh/KVIrc-5.0.0/src/kvilib/core/KviCString.cpp:1072
1072		m_len = str.m_len;
[Current thread is 1 (Thread 0x7ff8fd404440 (LWP 8963))]
(gdb) bt
#0  0x00007ff8fcf70096 in KviCString::KviCString(KviCString const&) (this=0x7ffc13d31290, str=...) at /home/wssh/KVIrc-5.0.0/src/kvilib/core/KviCString.cpp:1072
#1  0x0000555bfe387b8c in KviIrcMessage::trailingString() (this=0x7ffc13d313d0) at /home/wssh/KVIrc-5.0.0/src/kvirc/sparser/KviIrcMessage.h:127
#2  0x0000555bfe376514 in KviIrcServerParser::parseLiteralPrivmsg(KviIrcMessage*) (this=0x555c008c2460, msg=0x7ffc13d313d0) at /home/wssh/KVIrc-5.0.0/src/kvirc/sparser/KviIrcServerParser_literalHandlers.cpp:876
#3  0x0000555bfe39f4f1 in KviIrcServerParser::parseMessage(char const*, KviIrcConnection*) (this=0x555c008c2460, message=0x555c00e17f40 ":C PRIVMSG    ", pConnection=0x555c00d94640) at /home/wssh/KVIrc-5.0.0/src/kvirc/sparser/KviIrcServerParser.cpp:111
#4  0x0000555bfe48f0ba in KviIrcConnection::incomingMessage(char const*) (this=0x555c00d94640, pcMessage=0x555c00e17f40 ":C PRIVMSG    ") at /home/wssh/KVIrc-5.0.0/src/kvirc/kernel/KviIrcConnection.cpp:1938
#5  0x0000555bfe4a932a in KviIrcLink::processData(char*, int) (this=0x555c00e1a330, buffer=0x7ffc13d31580 ":C PRIVMSG    \n", iLen=15) at /home/wssh/KVIrc-5.0.0/src/kvirc/kernel/KviIrcLink.cpp:252
#6  0x0000555bfe4ae5a7 in KviIrcSocket::readData(int) (this=0x555c00e12a90) at /home/wssh/KVIrc-5.0.0/src/kvirc/kernel/KviIrcSocket.cpp:1526
#7  0x0000555bfe4e3344 in KviIrcSocket::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=0x555c00e12a90, _c=QMetaObject::InvokeMetaMethod, _id=2, _a=0x7ffc13d31af0) at /home/wssh/KVIrc-5.0.0/release/src/kvirc/kvirc_autogen/7GB2EGQPHR/moc_KviIrcSocket.cpp:94
#8  0x00007ff8fac49645 in QMetaObject::activate(QObject*, int, int, void**) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007ff8fac55ac8 in QSocketNotifier::activated(int, QSocketNotifier::QPrivateSignal) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#10 0x00007ff8fac55e92 in QSocketNotifier::event(QEvent*) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ff8fbd2483c in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#12 0x00007ff8fbd2c104 in QApplication::notify(QObject*, QEvent*) () at /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#13 0x00007ff8fac1a9c8 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#14 0x00007ff8fac745dd in  () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#15 0x00007ff8f8117417 in g_main_context_dispatch () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#16 0x00007ff8f8117650 in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#17 0x00007ff8f81176dc in g_main_context_iteration () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#18 0x00007ff8fac7397f in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#19 0x00007ff8fac189fa in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#20 0x00007ff8fac21aa4 in QCoreApplication::exec() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#21 0x0000555bfe4b59f9 in main(int, char**) (argc=1, argv=0x7ffc13d32178) at /home/wssh/KVIrc-5.0.0/src/kvirc/kernel/KviMain.cpp:488
(gdb) 

```

**Comment by AndrioCelos** [comment]
Thoughts on replacing the four different '`KviIrcMessage` trailing parameter' getters with only one that does the bounds check properly?

**Comment by ghost** [comment]
https://tools.ietf.org/html/rfc2812
Doesn't `QString` provide the correct conventions?
write the getters and setters to a private field? encapsulation?
As I'm only fuzzing the client because It was running on my computer and I'm paranoid.
Figured the Devs of the app would have the answer. Besides that, the string in question to cause the crash must only be sent by the server, and it was (because fuzzing) a user, through their client will never be able to cause this crash, correct me if I'm wrong. I do appologize for crapping in everyone's cereal. 

Correct me if I'm wrong, I was just fuzzing for fun. Perhaps someone knows more?
Perhaps someone else has something to add.

**Comment by ctrlaltca** [comment]
> Doesn't `QString` provide the correct conventions?

At that point in the parsing, the encoding of the message is not determined yet: while the server-specific part of the message is usually plain ASCII, the user-generated part of the message can use any encoding.That's why the message is not parsed into a QString yet, but a specific KviCString class is used instead.

Looking at the crash, it seems like changing `trailingString()` to `safeTrailingString()` at KviIrcServerParser_literalHandlers.cpp:876 should fix the issue.

**Comment by DarthGandalf** [comment]
@dem-ionz well, while it's not very easily triggerable by someone who is not an irc op, it still needs to be fixed. Thanks for finding and reporting it!

@ctrlaltca is there any reason to use the unsafe versions of these methods?

**Comment by ghost** [comment]
Will be happy to provide more fuzzing after the fix. :)

**Comment by wodim** [comment]
You can start right now, this client is in serious need of fuzzing I'd say.

**Comment by ctrlaltca** [comment]
@DarthGandalf before 7d08b0c3d7e1d4450bcb8a96abc7518c5a115917 the `m_pParams` vector used to be a `KviPointerList<KviCString> *`.
If the list is empty, calling `KviPointerList::last()` would return 0 (nullptr); in this case `trailingString()` would return 0 (nullptr), while `safeTrailingString()` would return a default-constructed empty `KviCString`.
A simple `if()` is in place just after the `trailingString()` call to check if the values is 0 and skip the code using it. This avoided the need of creating an useless empty `KviCString`.

From 7d08b0c3d7e1d4450bcb8a96abc7518c5a115917 on, that `KviPointerList::last()` call has been changed to a `std::vector::back()` call; unfortunately calling `back()` on an empty `std::vector` is undefined behavior.

**Comment by ctrlaltca** [comment]
Also, I'm scared at all the other possible bugs that have probably been introduced by replacing the 20-years old working code with the "most similar" std equivalent. 😨

**Comment by ghost** [comment]
Well, i'll fuzz the holy bytes out of it, once we can confirm and push a fix for the current bug to master. :)

**Comment by DarthGandalf** [comment]
> Also, I'm scared at all the other possible bugs that have probably been introduced by replacing the 20-years old working code with the "most similar" std equivalent. 

I wonder how many bugs were fixed at the same time :) But yes, ideally, such refactor is better done under a good test coverage. Which we also miss.

**Comment by ghost** [comment]
By modifying the code mentioned above, changing lines 876-877 in src/kvirc/sparser/KviIrcServerParser_literalHandlers.cpp to the following resolves this issue in our testing. I'm not sure if the fix is logical for the project as a whole, but for whatever it is actually worth this does fix it so I can continue fuzzing.

```
	KviCString pTrailing = msg->safeTrailingString();
	if(!strcmp(pTrailing,""))
```

**Comment by ghost** [comment]
FYI - You're going to want to patch 1257-1258 as well. The same fix lets us continue fuzzing.

**Comment by ghost** [comment]
Similar fixes can be done for 1006-1007 and 1449-1450 as well, updating them to this.

```
                KviCString theMsg = msg->safeTrailingString(); // FIXME
                if(!strcmp(theMsg,""))
```

**Comment by ghost** [comment]
Well, after 30 minutes fuzzing with no more crashes - I can say this is FIXED. Done spamming now. Do what ye will.

**Comment by ghost** [comment]
Final file fix: #2455
